packages/context/src/inject.ts

```diff
@@ -331,19 +380,54 @@ export function describeInjectedArguments(
   return meta || [];
 }
 
-function resolveByTag(
+/**
+ * Inspect the target type
+ * @param injection
+ */
+function inspectTargetType(injection: Readonly<Injection>) {
+  let type = MetadataInspector.getDesignTypeForProperty(
+    injection.target,
+    injection.member!,
+  );
+  if (type) {
+    return type;
+  }
+  const designType = MetadataInspector.getDesignTypeForMethod(
+    injection.target,
+    injection.member!,
+  );
+  type =
+    designType.parameterTypes[
+      injection.methodDescriptorOrParameterIndex as number
+    ];
+  return type;
+}
+
+/**
+ * Resolve values by a binding filter function
+ * @param ctx Context object
+ * @param injection Injection information
+ * @param session Resolution session
+ */
+function resolveByFilter(
   ctx: Context,
   injection: Readonly<Injection>,
   session?: ResolutionSession,
 ) {
-  const tag: string | RegExp = injection.metadata!.tag;
-  const bindings = ctx.findByTag(tag);
+  const bindingFilter = injection.metadata!.bindingFilter;
+  const view = new ContextView(ctx, bindingFilter);
+  const watch = injection.metadata!.watch;
 
-  return resolveList(bindings, b => {
-    // We need to clone the session so that resolution of multiple bindings
-    // can be tracked in parallel
-    return b.getValue(ctx, ResolutionSession.fork(session));
-  });
+  const targetType = inspectTargetType(injection);
+  if (targetType === Function) {
+    if (watch !== false) view.watch();
```

Comment:
Please document the `watch` option. What are the use cases when the users would want to enable or disable the watch mode? How should they decide which mode to use?

Personally, I would remove the `watch` option for now and always call `view.watch()`, such behavior seems as most natural to me.